agent/consul/state/catalog.go

```diff
@@ -2179,3 +2366,194 @@ func checkSessionsTxn(tx *memdb.Txn, hc *structs.HealthCheck) ([]*sessionCheck,
 	}
 	return sessions, nil
 }
+
+// updateGatewayService associates services with gateways as specified in a terminating-gateway config entry
+func (s *Store) updateTerminatingGatewayServices(tx *memdb.Txn, idx uint64, conf structs.ConfigEntry, entMeta *structs.EnterpriseMeta) error {
+	entry, ok := conf.(*structs.TerminatingGatewayConfigEntry)
+	if !ok {
+		return fmt.Errorf("unexpected config entry type: %T", conf)
+	}
+
+	// Delete all associated with gateway first, to avoid keeping mappings that were removed
+	if _, err := tx.DeleteAll("terminating-gateway-services", "gateway", structs.NewServiceID(entry.Name, entMeta)); err != nil {
```

Comment:
Also wondering whether the delete all is going to cause extra churn in blocking queries in the case where the terminating gateways service mappings haven't changed but the config entry is being rewritten. I think the answer is that it wont because its all done in the same memdb transaction. Also we could detect that the config entry isn't different and just not insert it in the first place prior to making any attempt to update the mapping. We do this in a number of other places in the catalog code.

Reply:
Right, there shouldn't be extra churn because of the transaction. The write transaction is operating on a copy of the tree.

Added a check to see if the config entry's services list matches the one stored in the state store.

Reply:
Added a constant for the table name too.